instrumentation/elasticsearch/elasticsearch-rest-7.0/library/src/main/java/io/opentelemetry/instrumentation/elasticsearch/rest/v7_0/RestClientWrapper.java

```diff
@@ -0,0 +1,186 @@
+                    try (Scope scope = context.makeCurrent()) {
+                      response = (Response) method.invoke(target, args);
+                    } catch (Throwable exception) {
+                      throwable = exception;
+                    } finally {
+                      instrumenter.end(context, otelRequest, response, throwable);
+                    }
+
+                    return response;
+                  } else if ("performRequestAsync".equals(method.getName())
+                      && args.length == 2
+                      && args[0] instanceof Request
+                      && args[1] instanceof ResponseListener) {
+
+                    Request request = (Request) args[0];
+                    ResponseListener responseListener = (ResponseListener) args[1];
+                    Context parentContext = Context.current();
+                    ElasticsearchRestRequest otelRequest =
+                        ElasticsearchRestRequest.create(request.getMethod(), request.getEndpoint());
+                    if (!instrumenter.shouldStart(parentContext, otelRequest)) {
+                      return method.invoke(target, args);
+                    }
+
+                    Throwable throwable = null;
+                    Context context = instrumenter.start(parentContext, otelRequest);
+                    args[1] =
+                        new RestResponseListener(
+                            responseListener, parentContext, instrumenter, context, otelRequest);
+                    try (Scope scope = context.makeCurrent()) {
+                      return method.invoke(target, args);
+                    } catch (Throwable exception) {
+                      throwable = exception;
+                    } finally {
+                      if (throwable != null) {
+                        instrumenter.end(context, otelRequest, null, throwable);
+                      }
+                      // span ended in RestResponseListener
+                    }
+                  }
+
+                  // delegate to wrapped RestClient
+                  return method.invoke(target, args);
+                }))
+        .make()
+        .load(RestClient.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+        .getLoaded();
+  }
+
+  private static Field getTargetField(Class<?> clazz) {
+    return getProxyField(clazz, "target");
+  }
+
+  private static Field getInstrumenterField(Class<?> clazz) {
+    return getProxyField(clazz, "instrumenter");
+  }
+
+  private static Field getProxyField(Class<?> clazz, String fieldName) {
+    try {
+      return clazz.getDeclaredField(fieldName);
+    } catch (NoSuchFieldException exception) {
+      throw new IllegalStateException("Could not find proxy field", exception);
+    }
+  }
+
+  private static Function<RestClient, RestClient> getProxyFactory(Class<?> clazz) {
+    for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
```

Comment:
Here we expect that the class has a constructor that takes `Header[]` as second and `List` as third argument, these are the only 2 parameters that we need to fill with non default values. Alternatively we could skip the constructor call if there are too many changes in the constructors.